### ISSUE TITLE
Reuse the same state in multiple screens

### ISSUE THREAD

**Issue by ranhsd** [ReSwift/ReSwift#316]
Hi, I have some issue that I am facing with while using ReSwift in my project

Let's assume that I have the following navigation in my app:

Home --> Feed --> Item Details --> Item Owners --> Owner Items --> Item Details

Now the question is about using **Item Details** twice in the same navigation stack.

My item details subscribe to ItemDetailsState and the issue that I am facing with is the the second item details reducer (last in the navigation stack) override the first item details state. I know that I can just create 2 similar objects and use them but I really want to reuse the same thing because eventually I am doing it on the same screen.
I read a bit about it online and found that in React they solve it using Namespaces but I didn't found any docs about it in ReSwift world so that's why I decided to ask you this question here.

Thanks for any help!

**Comment by mjarvis** [comment]
One pattern you could use to resolve this would be to store multiple, keyed `Item Details`, and pass the key into the object subscribing to the state when its created.
For example, `[UUID: ItemDetailsState]`, then `Controller(identifier: UUID)`
Then any actions can include `sender: UUID` to ensure that the reducer modifies the correct state.

**Comment by danielmartinprieto** [comment]
Yes, I also suggest you to do that. And that’s pretty much the “namespacing” thing that Redux users do as well.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by ranhsd** [comment]
Then all actions which are dispached on this reducer must contain additional field of UUID? And in the reducer I need to compare this UUID to the action UUID and if they are not equal simply return the state ?

**Comment by mjarvis** [comment]
Heres a little helper you can wrap around your existing reducer, and a protocol for actions to conform to:
```swift
protocol IdentifiableAction: Action {
    var identifier: UUID { get }
}

identifiableActionReducer<T>(_ action: Action, state: [UUID: T]?, specificReducer: (Action, T?) -> T) -> [UUID: T] {
    var state = state ?? [:]

    guard let action = action as IdentifiableAction else {
        return state
    }
    
    state[action.identifier] = specificReducer(action, state[action.identifier])
    return state
}

// app reducer usage:

itemDetailsState: identifiableActionReducer(action, state: state?.itemDetailsState, specificReducer: itemDetailsReducer)

// action usage:

struct SomeItemDetailsAction: IdentifiableAction {
    let identifier: UUID
    let payload: String
}
```

**Comment by ranhsd** [comment]
Hi @mjarvis thanks a lot for that. But in my case my app reducer should contain 2 itemDetailsState or just one?

**Comment by mjarvis** [comment]
```swift
struct AppState {
    let itemDetailsState: [UUID: ItemsDetailState]
}

func appReducer(_ action: Action, state: AppState?) -> AppState {
    return AppState(
        itemDetailsState: identifiableActionReducer(...)
    )
}
```

**Comment by ranhsd** [comment]
Hi @mjarvis that's exactly what I did  but something is not working there and I get syntax errors. Here is how it looks like on my side: 

**My AppState**

```
struct AppState : StateType, HasNavigationState {
   let itemDetailsState : [UUID : ItemDetailsState]
}
```

My **AppReducer**

```
func appReducer(action: Action, state: AppState?) -> AppState {
    
        itemDetailsState : identifiableActionReducer(action, state: state?.itemDetailsState, specificReducer: itemDetailsReducer)    
}
```

In my **ItemDetailsViewController** I subscribe with: 

```
        let uuid = UUID()
        
        store.subscribe(self) { subscription in
            subscription.select({ state in state.itemDetailsState[uuid] })
        }
```

I get syntax error of: 

> Cannot convert value of type '(ItemDetailsState?, Action) -> ItemDetailsState' to expected argument type '(Action, ItemDetailsState?) -> ItemDetailsState'

Also, I get some syntax error in my ItemDetailsViewController

I really don't understand why I am getting it because it looks like everything is the same as you described..

**Comment by mjarvis** [comment]
it looks like the arguments are reversed in your definition of `itemDetailsReducer`

**Comment by ranhsd** [comment]
Hi @mjarvis yes they were reversed ... I didn't noticed that. Wanted to ask if you have some example project in Github about this topic. I am asking because there are a lot more things to understand when you want to integrate it to your project. For example: when I am trying to subscribe I get an error on unwrapping optional and so. Or maybe you have some gist or something similar.

Thanks a ton for your help!

**Comment by ranhsd** [comment]
Hi @mjarvis I play with it a bit and manage to make some of things works.. I notice that it's now really not override other view controllers like it did before. The only question that is open for me is the subscriptions:

1. Usually what I am doing to to subscribe to the state in the **viewDidLoad** with this approach I cannot do it because I must find a way to first create the reducer in the UUID place and then subscribe to it. So what do you think will be the best way to do it?

2. Unsubscribe - in the deinit I want to clear the reducer. I want to remove the UUID from the states dictionary and I wanted to understand how can I achieve this behaviour. I was thinking about creating a generic action for removing the identifier and then in the identifiableActionReducer check for this action and then just remove the row with the specific identifier from the dictionary. This is a good approach or there are any other better ways to achieve it? 

3. In my app in addition to the Item Details screen I have another screen which require this functionality. I've notice that the reducer will be triggered twice and will create state for each one of the identifiers and this is not good practice because I need only the one that I'm using it right now. 

Thanks a lot !
